### PR TITLE
04-GET-api-articles-article_id

### PR DESCRIPTION
**Pull request by seb-the-architect**
Hi. Thanks for looking at my pull request :)

### CODE REVIEW

**Comment by seb-the-architect** [comment]
I had to change the development data slightly. Unsupported UTF characters or something.
https://stackoverflow.com/q/52254412/7485608

File: __tests__/app.test.js [hunk 1]
```diff
@@ -1,27 +1,54 @@
...
-                expect.objectContaining({slug:expect.any(String), description:expect.any(String)});
-            }
-        });
+            .get("/api/topics")
+            .expect(200)
+            .then((response) => {
+                expect(response.body.topics.length === 3);
+                for (let eachObject of response.body.topics) {
+                    expect(eachObject)
+                        .toEqual(expect.objectContaining({ slug: expect.any(String), description: expect.any(String) }));
```
Comment (MrMcSnail) [inline __tests__/app.test.js:20]: Nice use of `.objectContaining`

File: controller/sql-query-controller.js [hunk 2]
```diff
@@ -6,4 +6,19 @@ exports.getAllTopics = async function(req, res)
     
     try { res.status(200).send({"topics": allTopics}) } 
     catch (error) { console.error(error); }
```
Comment (MrMcSnail) [inline controller/sql-query-controller.js]: The error needs to reach the client, so you'll need to make sure to respond to the client with an error. Console.error is an internal log.

File: controller/sql-query-controller.js [hunk 3]
```diff
@@ -6,4 +6,19 @@ exports.getAllTopics = async function(req, res)
...
+}
+
+exports.getArticle = async function(req, res)
+{
+    //If the id consists only of numbers (input validation / injection prevention)
+    if(/^[0-9]+$/.test(req.params.article_id)) 
+    { 
+        const article = await model.queryArticle(req.params.article_id)
+        try { res.status(200).send({"article": article}) } 
+        catch (error) { console.error(error); }
```
Comment (MrMcSnail) [inline controller/sql-query-controller.js]: as above

File: controller/sql-query-controller.js [hunk 4]
```diff
@@ -6,4 +6,19 @@ exports.getAllTopics = async function(req, res)
...
+    //If the id consists only of numbers (input validation / injection prevention)
+    if(/^[0-9]+$/.test(req.params.article_id)) 
+    { 
+        const article = await model.queryArticle(req.params.article_id)
+        try { res.status(200).send({"article": article}) } 
+        catch (error) { console.error(error); } 
+    }
+    else {
+        try { res.status(400).send({"error": "Bad request. Is the ID supplied correct?"}) } 
+        catch (error) { console.error(error); }
```
Comment (MrMcSnail) [inline controller/sql-query-controller.js]: see above

File: __tests__/app.test.js [hunk 5]
```diff
@@ -1,27 +1,54 @@
...
+                        title: expect.any(String),
+                        topic: expect.any(String),
+                        name: expect.any(String),
+                        created_at: expect.any(String),
+                        votes: expect.any(Number)
+                    }));
+            }
+            )
+    })
+
```
Comment (MrMcSnail) [inline __tests__/app.test.js:44]: What if the article ID is a number but does not exist in the DB?

**Review by MrMcSnail — CHANGES_REQUESTED** [review]
Please refer back to the notes on [Middleware](https://notes.northcoders.com/courses/js-back-end/middleware) and [Error Handling](https://notes.northcoders.com/courses/js-back-end/error-handling). YOur .catch block should invoke a middleware error handler in your app.

File: controller/sql-query-controller.js (hunk 2 above)
Comment (seb-the-architect) [reply controller/sql-query-controller.js]: Next() is now used.

**Pull request closed without merging**